### PR TITLE
Introduce dynamic_unique_cast.

### PR DESCRIPTION
**Pull request by bangerth**
We've needed this in a couple of places recently. Here is the implementation
that we can then use wherever necessary.

### CODE REVIEW

File: include/deal.II/base/std_cxx11/unique_ptr.h [hunk 1]
```diff
@@ -33,4 +33,70 @@ DEAL_II_NAMESPACE_OPEN
...
+ *     // also that in accordance with the semantics of std::unique_ptr,
+ *     // it was necessary to std::move the 'b' object, and indeed 'b'
+ *     // now no longer pointers to anything -- ownership has been
+ *     // transferred to 'd'!
+ * @endcode
+ *
+ * @note This function does not try to convert the `Deleter` objects stored
+ *   by `std::unique_ptr` objects. The function therefore only works if the
+ *   deleter objects are at their defaults.
+ */
```
Comment (masterleinad) [inline include/deal.II/base/std_cxx11/unique_ptr.h]: Can we have a `static_assert` for this? Something like
```
static_assert(std::is_same<std::unique_ptr<From>::deleter_type, std::default_delete<From>>::value, 
              "The std::unique_ptr given must have a default deleter!");
```
should work.

**Comment by masterleinad** [comment]
I have the necessary replacements (for `std::unique_ptr::release()`) prepared.

**Comment by bangerth** [comment]
Oh, I have the changes to `mapping_q.cc` ready as well. Let's have this merged first.

File: include/deal.II/base/std_cxx11/unique_ptr.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/base/std_cxx11/unique_ptr.h]: We don't need this -- the argument of the function is a `std::unique_ptr<From>`, which implies that the second argument is at its default value. We *know* that whenever the function is called the static assertion would be satisfied.

Comment (bangerth) [reply include/deal.II/base/std_cxx11/unique_ptr.h]: Maybe to be more clear: The compiler will reject a call to
```
  std::unique_ptr<B,MyDeleter> b;
  std::unique_ptr<D,SomeDeleter> y = dynamic_unique_cast<D>(std::move(b));
```
because the argument `b` does not match the type of the argument of `dynamic_unique_cast`.

Comment (masterleinad) [reply include/deal.II/base/std_cxx11/unique_ptr.h]: Thanks for clarifying! Indeed, 
```
<source>: In function 'void test()':
<source>:39:62: error: no matching function for call to 'dynamic_unique_cast<B>(std::remove_reference<std::unique_ptr<D, void (*)(D*)>&>::type)'
    std::unique_ptr<B> b = dynamic_unique_cast<B>(std::move(d));
                                                              ^
<source>:5:2: note: candidate: template<class To, class From> std::unique_ptr<_Tp> dynamic_unique_cast(std::unique_ptr<From>&&)
  dynamic_unique_cast (std::unique_ptr<From/*, Deleter*/> &&p)
  ^~~~~~~~~~~~~~~~~~~
<source>:5:2: note:   template argument deduction/substitution failed:
<source>:39:62: note:   mismatched types 'std::default_delete<From>' and 'void (*)(D*)'
    std::unique_ptr<B> b = dynamic_unique_cast<B>(std::move(d));
```
should be quite understandable.

Comment (bangerth) [reply include/deal.II/base/std_cxx11/unique_ptr.h]: Maybe as an additional data point: deleters are generally only useful for objects that have no destructors. Say, pointers to PETSc objects that need to be destroyed by hand via calling a special function. On the other hand, if you're using `dynamic_cast`, then that can only work if you have a class hierarchy, and these classes furthermore have virtual members. Chances are that you will have a virtual destructor (and if you don't, you should!), which implies that the default deleter is doing the right thing in calling this destructor when destroying the object.

In other words, for classes for which you want to use this function, you will rarely need anything other than the default deleter.

File: include/deal.II/base/std_cxx11/unique_ptr.h [hunk 2]
```diff
@@ -33,4 +33,70 @@ DEAL_II_NAMESPACE_OPEN
...
+ *     // Assume that we know for some reason that the object above must
+ *     // have created a D object but returned it as a std::unique_ptr<B>.
+ *     // In order to access the D functionality, we need to cast the
+ *     // pointer. Use the equivalent to dynamic_cast:
+ *     std::unique_ptr<D> d = dynamic_unique_cast<D>(std::move(b));
+ *
+ *     // If the object really was a D, then 'd' now points to it. Note
+ *     // also that in accordance with the semantics of std::unique_ptr,
+ *     // it was necessary to std::move the 'b' object, and indeed 'b'
+ *     // now no longer pointers to anything -- ownership has been
```
Comment (kronbichler) [inline include/deal.II/base/std_cxx11/unique_ptr.h]: pointers -> points

File: include/deal.II/base/std_cxx11/unique_ptr.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/base/std_cxx11/unique_ptr.h]: It would be nice (but it is certainly not a blocker) to mention `std::default_delete` by name here for clarity.

**Comment by bangerth** [comment]
I fixed the two editorial comments.

**Comment by drwells** [comment]
I did not realize that this was in `std_cxx11/unique_ptr.h` until this was merged. I really think this should go in another header since we will remove all of the old compatibility headers after the 9.0 release.

**Comment by bangerth** [comment]
Legitimate point. Move it to `namespace Utilities`?

**Comment by masterleinad** [comment]
Originally, I thought about `std_cxx14/memory.h`.

**Comment by drwells** [comment]
I believe the intent of the `std_*` folders is to provide compatibility work-arounds for older language versions. I do not think we should put our own stuff in there: Utilities seems like the right place to me.

**Comment by masterleinad** [comment]
> I believe the intent of the std_* folders is to provide compatibility work-arounds for older language versions. I do not think we should put our own stuff in there: Utilities seems like the right place to me.

Fine with me.

**Comment by bangerth** [comment]
Yes, I think of the `std_cxx??/` files as providing things we want to use from future standards at times when we don't want to require such compilers yet.
